nautobot_ssot/integrations/dna_center/utils/dna_center.py

```diff
@@ -0,0 +1,243 @@
+"""Utility functions for working with DNA Center."""
+
+import logging
+import re
+from typing import List, Tuple
+
+from dnacentersdk import api
+from dnacentersdk.exceptions import dnacentersdkException
+from netutils.constants import BASE_INTERFACES
+
+from nautobot_ssot.integrations.dna_center.constants import BASE_INTERFACE_MAP
+
+LOGGER = logging.getLogger(__name__)
+
+
+class DnaCenterClient:
+    """Client for handling all interactions with DNA Center."""
+
+    def __init__(
+        self, url: str, username: str, password: str, port: int = 443, verify: bool = True
+    ):  # pylint: disable=too-many-arguments
+        """Initialize instance of client."""
+        self.url = url
+        self.port = port
+        self.base_url = f"{self.url}:{self.port}"
+        self.username = username
+        self.password = password
+        self.verify = verify
+        self.conn = None
+
+    def connect(self):  # pylint: disable=inconsistent-return-statements
+        """Connect to Cisco DNA Center."""
+        try:
+            self.conn = api.DNACenterAPI(
+                base_url=self.base_url, username=self.username, password=self.password, verify=self.verify
+            )
+        except dnacentersdkException as err:
+            raise dnacentersdkException(f"Unable to connect to DNA Center: {err}") from err
+
+    def get_locations(self):
+        """Retrieve all location data from DNA Center.
+
+        Returns:
+            list: List of Locations (Sites) from DNAC.
+        """
+        locations, loc_data, loc_names = [], [], []
+        try:
+            total_num_sites = self.conn.sites.get_site_count()["response"]
+            offset = 1
+            while len(loc_data) < total_num_sites:
+                loc_data.extend(self.conn.sites.get_site(offset=offset)["response"])
+                offset = len(loc_data)
+            for _, item in enumerate(loc_data):
+                if item["siteNameHierarchy"] not in loc_names:
+                    loc_names.append(item["siteNameHierarchy"])
+                    locations.append(item)
+        except dnacentersdkException as err:
+            LOGGER.error("Unable to get site information from DNA Center. %s", err)
+        return locations
+
+    @staticmethod
+    def find_address_and_type(info: List[dict]):
+        """Find Site address and type from additionalInfo dict.
+
+        Args:
+            info (List[dict]): Site additionalInfo property from DNA Center.
+
+        Returns:
+            tuple: Tuple of Site address and type.
+        """
+        address = ""
+        site_type = ""
+        for element in info:
+            if element["nameSpace"] == "Location":
+                address = element["attributes"]["address"]
+                site_type = element["attributes"]["type"]
+        return (address, site_type)
```

Comment:
```suggestion
        for element in info:
            if element["nameSpace"] == "Location":
                return (element["attributes"]["address"], element["attributes"]["type"]) 
        return ("", "")
```

A bit more performant

Reply:
This can also be applied further down

Reply:
Changes have been made.